include/exadg/incompressible_navier_stokes/spatial_discretization/generalized_newtonian_model.cpp

```diff
@@ -0,0 +1,334 @@
+
+      add_generalized_newtonian_viscosity(viscosity, velocity_gradient);
+
+      // set the coefficients
+      Base::viscous_kernel->set_coefficient_face(face, q, viscosity);
+    }
+  }
+}
+
+template<int dim, typename Number>
+void
+GeneralizedNewtonianModel<dim, Number>::add_generalized_newtonian_viscosity(
+  scalar &       viscosity,
+  tensor const & velocity_gradient) const
+{
+  tensor symmetric_velocity_gradient = 0.5 * (velocity_gradient + transpose(velocity_gradient));
+
+  scalar shear_rate =
+    std::sqrt(2.0 * scalar_product(symmetric_velocity_gradient, symmetric_velocity_gradient));
+
+  scalar viscosity_factor;
+  compute_viscosity_factor(viscosity_factor, shear_rate);
+
+  viscosity += viscosity_factor * viscosity_margin;
+}
+
+template<int dim, typename Number>
+void
+GeneralizedNewtonianModel<dim, Number>::compute_viscosity_factor(scalar &       viscosity_factor,
+                                                                 scalar const & shear_rate) const
+{
+  switch(generalized_newtonian_model)
+  {
+    case GeneralizedNewtonianViscosityModel::Undefined:
+      AssertThrow(generalized_newtonian_model != GeneralizedNewtonianViscosityModel::Undefined,
+                  dealii::ExcMessage("parameter must be defined"));
+      break;
+    case GeneralizedNewtonianViscosityModel::GeneralizedCarreauYasuda:
+      generalized_carreau_yasuda_model(viscosity_factor, shear_rate);
+      break;
+    case GeneralizedNewtonianViscosityModel::Carreau:
+      carreau_model(viscosity_factor, shear_rate);
+      break;
+    case GeneralizedNewtonianViscosityModel::Cross:
+      cross_model(viscosity_factor, shear_rate);
+      break;
+    case GeneralizedNewtonianViscosityModel::SimplifiedCross:
+      simplified_cross_model(viscosity_factor, shear_rate);
+      break;
+    case GeneralizedNewtonianViscosityModel::PowerLaw:
+      power_law_model(viscosity_factor, shear_rate);
+      break;
+    default:
+      AssertThrow(
+        false, dealii::ExcMessage("This GeneralizedNewtonianViscosityModel is not implemented."));
+  }
+}
+
+template<int dim, typename Number>
+void
+GeneralizedNewtonianModel<dim, Number>::generalized_carreau_yasuda_model(
+  scalar &       viscosity_factor,
+  scalar const & shear_rate) const
+{
+  // eta = eta_oo + (eta_0 - eta_oo) * [k + (l * y)^a]^[(n-1)/a]
+  viscosity_factor = shear_rate * lambda;
+  viscosity_factor = pow(viscosity_factor, a);
+  viscosity_factor += kappa;
+  viscosity_factor = pow(viscosity_factor, static_cast<Number>((n - 1.0) / a));
+}
+
+template<int dim, typename Number>
+void
+GeneralizedNewtonianModel<dim, Number>::carreau_model(scalar &       viscosity_factor,
+                                                      scalar const & shear_rate) const
+{
+  // eta = eta_oo + (eta_0 - eta_oo) * [k + (l * y)^a]^[(n-1)/a]
+  // with k = 1 and a = 2
+  // eta = eta_oo + (eta_0 - eta_oo) * [1 + l^2 * y^2]^[(n-1)/2]
+  viscosity_factor = shear_rate * shear_rate * (lambda * lambda);
```

Comment:
As written above, I would implement it differently. Nevertheless, here one has unnecessarily many multiplications. The following code only needs 2 instead of 3 multiplications:

```C++
viscosity_factor = shear_rate * lambda;
viscosity_factor *= viscosity_factor;
```